src/goto-instrument/code_contracts.cpp

```diff
@@ -370,25 +402,74 @@ void code_contractst::add_contract_check(
   dest.destructive_insert(dest.instructions.begin(), check);
 }
 
-void code_contractst::operator()()
+bool code_contractst::replace_calls(
+  const std::list<std::string> &funs_to_replace)
 {
-  Forall_goto_functions(it, goto_functions)
-    code_contracts(it->second);
-
-  goto_functionst::function_mapt::iterator i_it=
-    goto_functions.function_map.find(INITIALIZE_FUNCTION);
-  assert(i_it!=goto_functions.function_map.end());
+  bool fail = false;
+  Forall_goto_functions(fit, goto_functions)
+  {
+    Forall_goto_program_instructions(it, fit->second.body)
+    {
+      if(it->is_function_call())
+      {
+        const code_function_callt &call = it->get_function_call();
+
+        // TODO we don't handle function pointers
+        if(call.function().id() != ID_symbol)
+          continue;
+
+        const irep_idt &fun_name =
+          to_symbol_expr(call.function()).get_identifier();
+        auto found = std::find(
+          funs_to_replace.begin(), funs_to_replace.end(), id2string(fun_name));
+        if(found == funs_to_replace.end())
+          continue;
+
+        if(!has_contract(fun_name))
+        {
+          log.error() << "Function '" << fun_name
+                      << "' does not have a contract; "
+                      << "not replacing call with contract." << messaget::eom;
+          fail = true;
+          continue;
+        }
+
+        fail |= apply_contract(fit->second.body, it);
+      }
+    }
+  }
 
-  for(const auto &contract : summarized)
-    add_contract_check(contract, i_it->second.body);
+  if(fail)
+    return true;
 
-  // remove skips
-  remove_skip(i_it->second.body);
+  Forall_goto_functions(fit, goto_functions)
+  {
+    remove_skip(fit->second.body);
+  }
 
   goto_functions.update();
+
+  return false;
 }
 
-void code_contracts(goto_modelt &goto_model)
+bool code_contractst::replace_calls()
 {
-  code_contractst(goto_model.symbol_table, goto_model.goto_functions)();
+  std::list<std::string> funs_to_replace;
```

Comment:
Is there a good reason for this to be a `std::list`? It has a pretty big memory and performance overhead, unless you need stable iterators it’s probably better to just use a `vector`. (Same with the other instances of `list`).